### ISSUE TITLE
Heavy loading of CPU

### ISSUE THREAD

**Issue by lindeloev** [ueffel/Keypirinha-Plugin-Kill#15]
I reported a heavy CPU load [here](https://github.com/Keypirinha/Keypirinha/issues/334) when launching KP as non-admin, and it turned out to be specific to the `Kill` plugin. I'm happy to help debugging.

KP version: 2.21, installed via choco install, Windows 10 64 bit. The `Kill` version is whatever is installed when running the `PackageControl install` command. 

Here's the log for the startup and where it stops, when Kill is installed:

```
2018-10-31 11:57:04.413 [i] Loaded 586 items from history file.
2018-10-31 11:57:04.423 [i] Loaded 1138 items from catalog file.
2018-10-31 11:57:04.460 [i] Init done in 0.2694s, including 0.1681s (62%) for the Python interpreter.
2018-10-31 11:57:04.460 [i] Plugin loaded: Internal.About
2018-10-31 11:57:04.502 [i] Plugin loaded: Apps.CustomCmds (instance #1)
2018-10-31 11:57:04.502 [i] Plugin loaded: Apps.Desktop (instance #1)
2018-10-31 11:57:04.502 [i] Plugin loaded: Apps.EnvPath (instance #1)
2018-10-31 11:57:04.502 [i] Plugin loaded: Apps.ExtraPaths (instance #1)
2018-10-31 11:57:04.502 [i] Plugin loaded: Apps.StartMenu (instance #1)
2018-10-31 11:57:04.513 [i] Plugin loaded: Bookmarks.Bookmarks (instance #1)
2018-10-31 11:57:04.525 [i] Plugin loaded: Calc.Calc (instance #1)
2018-10-31 11:57:04.527 [i] Plugin loaded: ControlPanel.ControlPanel (instance #1)
2018-10-31 11:57:04.537 [i] Plugin loaded: Currency.Currency (instance #1)
2018-10-31 11:57:04.541 [i] Plugin loaded: Env.Env (instance #1)
2018-10-31 11:57:04.546 [i] Plugin loaded: Everything.Everything (instance #1)
2018-10-31 11:57:04.549 [i] Plugin loaded: FileBrowser.FileBrowser (instance #1)
2018-10-31 11:57:04.562 [i] Plugin loaded: FilesCatalog.FilesCatalog (instance #1)
2018-10-31 11:57:04.570 [i] Plugin loaded: FileZilla.FileZilla (instance #1)
2018-10-31 11:57:04.573 [i] Plugin loaded: GoogleTranslate.GoogleTranslate (instance #1)
------------------------ STOPS HERE WITH KILL INSTALLED ---------------------------
2018-10-31 11:57:04.638 [i] Plugin loaded: Kill.Kill (instance #1)
2018-10-31 11:57:04.644 [i] Plugin loaded: PackageControl.PackageControl (instance #1)
2018-10-31 11:57:04.646 [i] Plugin loaded: PuTTY.PuTTY (instance #1)
2018-10-31 11:57:04.650 [i] Plugin loaded: RegBrowser.RegBrowser (instance #1)
2018-10-31 11:57:04.654 [i] Plugin loaded: String.String (instance #1)
2018-10-31 11:57:04.656 [i] Plugin loaded: SystemCommands.SystemCommands (instance #1)
2018-10-31 11:57:04.659 [i] Plugin loaded: TaskSwitcher.TaskSwitcher (instance #1)
2018-10-31 11:57:04.662 [i] Plugin loaded: Units.Units (instance #1)
2018-10-31 11:57:04.664 [i] Plugin loaded: URL.URL (instance #1)
2018-10-31 11:57:04.667 [i] Plugin loaded: WebSearch.WebSearch (instance #1)
2018-10-31 11:57:04.672 [i] Plugin loaded: WebSuggest.WebSuggest (instance #1)
2018-10-31 11:57:04.675 [i] Plugin loaded: Winamp.Winamp (instance #1)
2018-10-31 11:57:04.677 [i] Plugin loaded: WinSCP.WinSCP (instance #1)
2018-10-31 11:57:04.757 [i] Apps.Desktop: Cataloged 4 items in 0.0 seconds
2018-10-31 11:57:04.775 [i] Apps.StartMenu: Cataloged 158 items in 0.1 seconds
2018-10-31 11:57:04.794 [i] PackageControl.PackageControl: Package list loaded from file cache 'ueffel's Package Repository' (28 packages)
2018-10-31 11:57:04.810 [i] Apps.EnvPath: Cataloged 744 items in 0.1 seconds
2018-10-31 11:57:05.148 [i] Bookmarks.Bookmarks: Referenced 42 bookmarks
2018-10-31 11:57:11.747 [i] PackageControl.PackageControl: Removed package: Keypirinha-Plugin-Kill
2018-10-31 11:57:13.245 [i] Unloaded package: Kill
2018-10-31 11:57:21.849 [i] PackageControl.PackageControl: Installed package 'Keypirinha-WindowsApps'
2018-10-31 11:57:23.399 [i] Plugin loaded: WindowsApps.WindowsApps (instance #1)
2018-10-31 11:57:25.561 [i] WindowsApps.WindowsApps: Cataloged 78 items in 2.2 seconds
```

**Comment by ueffel** [comment]
I have written some debug outputs in a new "debug version" of the package:
[Kill.zip](https://github.com/ueffel/Keypirinha-Plugin-Kill/files/2534180/Kill.zip)
 Can you download the file, rename it to `Kill.keypirinha-package` and replace your existing package file. 
(If you enter `import keypirinha as kp,os; os.system('explorer ' + kp.installed_package_dir())` in the input box in the lower part of the keypirinha console window, it should open the windows explorer at the right location)

Then restart keypirinha and post the contents of the log here please.

**Comment by lindeloev** [comment]
Done! It prints the following::

    2018-10-31 13:38:39.953 [i] Python: Kill: before comtypes import

**Comment by ueffel** [comment]
So there is no Output like `Python: Kill: before CommandLineToArgvW definition` ?
Can you post the whole content of the console after start? like this:
```
14:02:37.137 Keypirinha 2.21 (d8c4b9b) for x64
14:02:37.137 System: WinNT-x64 6.3.9600-ws-0x0100
14:02:37.137 Portable mode
14:02:37.137 Keyboard layout: 00000407
14:02:37.137 Monitor #1: Name[\\.\DISPLAY1] Rect[0, 0, 2560, 1440] DpiScale[1.00] PRIMARY
14:02:37.137 Monitor #2: Name[\\.\DISPLAY2] Rect[2560, 360, 4480, 1440] DpiScale[1.00]
14:02:37.137 Official packages: R:\tmp\Keypirinha\default\Packages
14:02:37.137 Profile dir: R:\tmp\Keypirinha\portable\Profile
14:02:37.137 Local dir: R:\tmp\Keypirinha\portable\Local
14:02:37.154 Python 3.6.3 (default, Oct  9 2017, 08:05:04) [MSC v.1900 64 bit (AMD64)]
14:02:37.409 Loaded 11 items from history file.
14:02:37.434 Loaded 3082 items from catalog file.
14:02:37.446 Init done in 0.5271s, including 0.2491s (47%) for the Python interpreter.
14:02:37.446 Plugin loaded: Internal.About
14:02:37.468 Plugin loaded: Apps.CustomCmds (instance #1)
14:02:37.468 Plugin loaded: Apps.Desktop (instance #1)
14:02:37.468 Plugin loaded: Apps.EnvPath (instance #1)
14:02:37.468 Plugin loaded: Apps.ExtraPaths (instance #1)
14:02:37.468 Plugin loaded: Apps.StartMenu (instance #1)
14:02:37.497 Plugin loaded: Bookmarks.Bookmarks (instance #1)
14:02:37.506 Plugin loaded: Calc.Calc (instance #1)
14:02:37.508 Plugin loaded: ControlPanel.ControlPanel (instance #1)
14:02:37.517 Plugin loaded: Currency.Currency (instance #1)
14:02:37.519 Plugin loaded: Env.Env (instance #1)
14:02:37.524 Plugin loaded: Everything.Everything (instance #1)
14:02:37.526 Plugin loaded: FileBrowser.FileBrowser (instance #1)
14:02:37.537 Plugin loaded: FilesCatalog.FilesCatalog (instance #1)
14:02:37.560 Plugin loaded: FileZilla.FileZilla (instance #1)
14:02:37.563 Plugin loaded: GoogleTranslate.GoogleTranslate (instance #1)
14:02:37.567 Python: Kill: before comtypes import
14:02:37.616 Python: Kill: before CommandLineToArgvW definition
14:02:37.625 Kill.Kill(50): DEBUG: __init__
14:02:37.625 Plugin loaded: Kill.Kill (instance #1)
14:02:37.630 Plugin loaded: PackageControl.PackageControl (instance #1)
14:02:37.633 Plugin loaded: PuTTY.PuTTY (instance #1)
14:02:37.635 Plugin loaded: RegBrowser.RegBrowser (instance #1)
14:02:37.639 Plugin loaded: String.String (instance #1)
14:02:37.641 Plugin loaded: TaskSwitcher.TaskSwitcher (instance #1)
14:02:37.644 Plugin loaded: URL.URL (instance #1)
14:02:37.646 Plugin loaded: WebSearch.WebSearch (instance #1)
14:02:37.649 Plugin loaded: WebSuggest.WebSuggest (instance #1)
14:02:37.652 Plugin loaded: Winamp.Winamp (instance #1)
14:02:37.654 Plugin loaded: WinSCP.WinSCP (instance #1)
14:02:37.707 Kill.Kill(99): DEBUG: on_start entry
14:02:37.717 Apps.Desktop: Cataloged 131 items in 0.0 seconds
14:02:37.728 Kill.Kill(69): DEBUG: Reading config
14:02:37.755 Apps.StartMenu: Cataloged 505 items in 0.1 seconds
14:02:37.757 PackageControl.PackageControl: Package list loaded from file cache 'ueffel's Package Repository' (28 packages)
14:02:37.760 Kill.Kill(87): DEBUG: default_action = kill_by_id
14:02:37.763 Kill.Kill(90): DEBUG: hide_background = False
14:02:37.765 Kill.Kill(93): DEBUG: item_label = Kill:
14:02:37.767 Kill.Kill(94): DEBUG: Reading config done
14:02:37.778 Kill.Kill(144): DEBUG: on_start before icon load
14:02:37.781 Kill.Kill(146): DEBUG: on_start done
14:02:37.921 Apps.EnvPath: Cataloged 2385 items in 0.2 seconds
14:02:44.417 Bookmarks.Bookmarks: Referenced 3865 bookmarks
```

**Comment by lindeloev** [comment]
As far as I can see, the line I sent you was the only output from the Kill module:

```
2018-10-31 13:38:39.704 [i] Loaded 598 items from history file.
2018-10-31 13:38:39.720 [i] Loaded 1219 items from catalog file.
2018-10-31 13:38:39.811 [i] Init done in 0.5204s, including 0.3282s (63%) for the Python interpreter.
2018-10-31 13:38:39.811 [i] Plugin loaded: Internal.About
2018-10-31 13:38:39.866 [i] Plugin loaded: Apps.CustomCmds (instance #1)
2018-10-31 13:38:39.866 [i] Plugin loaded: Apps.Desktop (instance #1)
2018-10-31 13:38:39.866 [i] Plugin loaded: Apps.EnvPath (instance #1)
2018-10-31 13:38:39.866 [i] Plugin loaded: Apps.ExtraPaths (instance #1)
2018-10-31 13:38:39.866 [i] Plugin loaded: Apps.StartMenu (instance #1)
2018-10-31 13:38:39.878 [i] Plugin loaded: Bookmarks.Bookmarks (instance #1)
2018-10-31 13:38:39.889 [i] Plugin loaded: Calc.Calc (instance #1)
2018-10-31 13:38:39.891 [i] Plugin loaded: ControlPanel.ControlPanel (instance #1)
2018-10-31 13:38:39.902 [i] Plugin loaded: Currency.Currency (instance #1)
2018-10-31 13:38:39.904 [i] Plugin loaded: Env.Env (instance #1)
2018-10-31 13:38:39.910 [i] Plugin loaded: Everything.Everything (instance #1)
2018-10-31 13:38:39.913 [i] Plugin loaded: FileBrowser.FileBrowser (instance #1)
2018-10-31 13:38:39.937 [i] Plugin loaded: FilesCatalog.FilesCatalog (instance #1)
2018-10-31 13:38:39.945 [i] Plugin loaded: FileZilla.FileZilla (instance #1)
2018-10-31 13:38:39.949 [i] Plugin loaded: GoogleTranslate.GoogleTranslate (instance #1)
2018-10-31 13:38:39.953 [i] Python: Kill: before comtypes import
```

I can't post the output of the console since it never launches the GUI when running as non-admin.

**Comment by ueffel** [comment]
Alright. There seems to be a problem with the import mechanism of the embedded python interpreter. It locks the whole application for some reason. This needs further investigation but I don't think its limited to the Kill package.

**Comment by ueffel** [comment]
Ok I can reproduce the issue, when I change my permissions of the `Keypirinha\python\site` directory and remove the right to write in a portable installation. It seems read and execute permission are not sufficient.
@polyvertex can you help me out here?

**Comment by ueffel** [comment]
@lindeloev Running Keypirinha as Administrator seems to be the culprit. This creates the directory `Keypirinha\python\site\comtypes\gen`. If you delete that directory from your installation and run keypirinha as user (not administrator) everything should work fine. But running once as administrator create this directory again.
I did not fully understand this weird behaviour of comtypes but that could be your work around for now.

**Comment by lindeloev** [comment]
I can confim that KP now work with the Kill module and everything as
regular user! Thanks so much for figuring this out.

Best,
Jonas
______________________________

Jonas Kristoffer Lindeløv, M.Sc., Ph.D.
Assistant Professor in Cognitive Neuroscience and Neuropsychology
Profile at Aalborg University <http://personprofil.aau.dk/117060>, Scientific
blog <http://lindeloev.net/>

**Comment by polyvertex** [comment]
This should be fixed in Keypirinha v2.22 which has just been released.
Thanks for the feedback and for investigating!

**Comment by ueffel** [comment]
Fixed with new version of Keypirinha v2.22.
